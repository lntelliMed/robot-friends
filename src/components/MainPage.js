import React, { Component } from "react";

import SearchBox from "./SearchBox";
import RobotList from "./RobotList";
import Header from "./Header";
import Scroll from "./Scroll";
import ErrorBoundary from "./ErrorBoundary";

class MainPage extends Component {
  componentDidMount() {
    this.props.onInitRobots();
  }

  filterRobots = () => {
    const { robots, searchField } = this.props;
    return robots.filter(robot => {
      return robot.name.toLowerCase().includes(searchField.toLowerCase());
    });
  };

  render() {
    const {
      robots,
      searchField,
      onSearchFieldChange,
      isPending,
      error
    } = this.props;
    let robotList = "";

    if (isPending) {
      robotList = <h2>Loading robots! Please wait...</h2>;
    } else if (error) {
      robotList = <h2>Something went wrong!</h2>;
    }

    if (robots.length) {
      const filteredRobots = this.filterRobots(robots, searchField);
      robotList = <RobotList robots={filteredRobots} />;
    }

    return (
      <div className="tc">
        <Header />
        <SearchBox searchChangeHandler={onSearchFieldChange} />
        <Scroll>
          <ErrorBoundary>{robotList}</ErrorBoundary>
        </Scroll>
      </div>
    );
  }
}

export default MainPage;
