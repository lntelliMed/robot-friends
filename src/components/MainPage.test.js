import React from "react";
import { shallow } from "enzyme";
import MainPage from "./MainPage";

let wrapper;

beforeEach(() => {
  const mockProps = {
    onInitRobots: jest.fn(),
    onSearchFieldChange: jest.fn(),
    robots: [],
    isPending: false,
    error: null,
    searchField: ""
  };

  wrapper = shallow(<MainPage {...mockProps} />);
});

it("MainPage renders without crashing", () => {
  expect(wrapper).toMatchSnapshot();
});

it("works correctly even if Robots is initially an empty array", () => {
  const mockProps = {
    onInitRobots: jest.fn(),
    onSearchFieldChange: jest.fn(),
    robots: [],
    isPending: false,
    error: null,
    searchField: "a"
  };

  wrapper = shallow(<MainPage {...mockProps} />);
  expect(
    wrapper.instance().filterRobots(mockProps.robots, mockProps.searchField)
  ).toEqual([]);
});

it("filters Robots correctly whenever a match is found", () => {
  const filteredRobots = [
    {
      id: 1,
      name: "Leanne Graham",
      username: "Bret",
      email: "Sincere@april.biz"
    }
  ];

  const mockProps = {
    onInitRobots: jest.fn(),
    onSearchFieldChange: jest.fn(),
    robots: [
      {
        id: 1,
        name: "Leanne Graham",
        username: "Bret",
        email: "Sincere@april.biz"
      }
    ],
    isPending: false,
    error: null,
    searchField: "Leanne"
  };

  wrapper = shallow(<MainPage {...mockProps} />);
  expect(
    wrapper.instance().filterRobots(mockProps.robots, mockProps.searchField)
  ).toEqual(filteredRobots);
});

it("works correctly even if no match is found", () => {
  const filteredRobots = [
    {
      id: 1,
      name: "Leanne Graham",
      username: "Bret",
      email: "Sincere@april.biz"
    }
  ];

  const mockProps = {
    onInitRobots: jest.fn(),
    onSearchFieldChange: jest.fn(),
    robots: [
      {
        id: 1,
        name: "Leanne Graham",
        username: "Bret",
        email: "Sincere@april.biz"
      }
    ],
    isPending: false,
    error: null,
    searchField: "Xavier"
  };

  wrapper = shallow(<MainPage {...mockProps} />);
  expect(
    wrapper.instance().filterRobots(mockProps.robots, mockProps.searchField)
  ).toEqual([]);
});
