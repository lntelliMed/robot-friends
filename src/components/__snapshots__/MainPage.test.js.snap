// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainPage
    error={null}
    isPending={false}
    onInitRobots={[Function]}
    onSearchFieldChange={[Function]}
    robots={Array []}
    searchField=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header />,
        <searchBox
          searchChangeHandler={[Function]}
/>,
        <scroll>
          <ErrorBoundary />
</scroll>,
      ],
      "className": "tc",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "searchChangeHandler": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ErrorBoundary />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header />,
          <searchBox
            searchChangeHandler={[Function]}
/>,
          <scroll>
            <ErrorBoundary />
</scroll>,
        ],
        "className": "tc",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "searchChangeHandler": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ErrorBoundary />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
